lls/cache.c

```diff
@@ -351,51 +350,119 @@ xmit_icmp_ping_reply(struct gatekeeper_if *iface, struct rte_mbuf *pkt)
 			RTE_ETHER_TYPE_IPV4);
 	}
 
-	/*
-	 * Set-up IPv4 header.
-	 *
-	 * According to RFC 792 page 13: to form an echo reply
-	 * message, the source and destination addresses are simply reversed,
-	 * the type code changed to 0, and the checksum recomputed.
-	 */
+	/* Set-up IPv4 header. */
 	icmp_ipv4 = (struct rte_ipv4_hdr *)pkt_out_skip_l2(iface, icmp_eth);
+	icmph = (struct rte_icmp_hdr *)ipv4_skip_exthdr(icmp_ipv4);
 
 	icmp_ipv4->time_to_live = IP_DEFTTL;
-	ip_addr_tmp = icmp_ipv4->src_addr;
-	icmp_ipv4->src_addr = icmp_ipv4->dst_addr;
-	icmp_ipv4->dst_addr = ip_addr_tmp;
+	switch (icmph->icmp_type) {
+	case ICMP_ECHO_REQUEST_TYPE: {
+		/*
+		 * According to RFC 792 page 13: to form an echo reply
+		 * message, the source and destination addresses are simply
+		 * reversed, the type code changed to 0, and the checksum
+		 * recomputed.
+		 */
+		rte_be32_t ip_addr_tmp = icmp_ipv4->src_addr;
+		icmp_ipv4->src_addr = icmp_ipv4->dst_addr;
+		icmp_ipv4->dst_addr = ip_addr_tmp;
+
+		/*
+		 * According to RFC 792 page 14:
+		 *
+		 * (1) the data received in the echo message must be returned in
+		 * the echo reply message.
+		 *
+		 * (2) the identifier and sequence number may be used by
+		 * the echo sender to aid in matching the replies with
+		 * the echo requests.
+		 *
+		 * So, we keep these fields unmodified.
+		 */
+		icmph->icmp_type = ICMP_ECHO_REPLY_TYPE;
+		icmph->icmp_code = ICMP_ECHO_REPLY_CODE;
+
+		break;
+	}
+	case ICMP_MASK_REQUEST_TYPE: {
+		uint32_t pkt_size;
+		struct in_addr *ip4_mask;
+
+		/*
+		 * According to RFC 3021: the 255.255.255.255 IP broadcast
+		 * address MUST be used for broadcast Address Mask Replies
+		 * in point-to-point links with 31-bit subnet masks.
+		 * In other cases, the form {<Network-number>, -1} should be
+		 * used.
+		 */
+		icmp_ipv4->src_addr = icmp_ipv4->dst_addr;
+		if (iface->ip4_addr_plen >= 31)
+			icmp_ipv4->dst_addr = (uint32_t)-1;
+		else {
+			icmp_ipv4->dst_addr = (iface->ip4_addr.s_addr |
+					(~iface->ip4_mask.s_addr));
+		}
+
+		pkt_size = sizeof(*icmp_eth) + iface->l2_len_out +
+			ipv4_hdr_len(icmp_ipv4) + sizeof(*icmph) +
+			sizeof(struct in_addr);
+		if (pkt->pkt_len < pkt_size) {
+			char *data = rte_pktmbuf_append(pkt, pkt_size -
+					pkt->pkt_len);
+			if (data == NULL) {
+				LLS_LOG(WARNING,
+						"Could not append %d bytes of data to the ICMP mask reply packet\n",
+						pkt_size -
+						pkt->pkt_len);
+				return false;
+			}
+		} else if (pkt->pkt_len > pkt_size) {
+			int ret = rte_pktmbuf_trim(pkt, pkt->pkt_len -
+					pkt_size);
+			if (ret != 0) {
+				LLS_LOG(WARNING,
+						"Could not remove %d bytes of data at the end of the ICMP mask reply packet\n",
+						pkt->pkt_len -
+						pkt_size);
+				return false;
+			}
+		}
+
+		icmp_ipv4->total_length = htonl(ipv4_hdr_len(icmp_ipv4) +
```

Comment:
Replace `htonl()` with `rte_cpu_to_be_16()` to match the type of `icmp_ipv4->total_length` and to avoid a bug due to the mix up of 16-bit and 32-bit sizes; `htonl()` is for 32-bit intergers.

Reply:
The expression inside `rte_cpu_to_be_16()` can be simplified with `pkt_size - iface->l2_len_out`.